sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsContainer.java

```diff
@@ -0,0 +1,201 @@
+   */
+  public MetricsContainer(String stepName) {
+    this.stepName = stepName;
+  }
+
+  /**
+   * Set the {@link MetricsContainer} for the current thread.
+   */
+  public static void setMetricsContainer(MetricsContainer container) {
+    CONTAINER_FOR_THREAD.set(container);
+  }
+
+  /**
+   * Clear the {@link MetricsContainer} for the current thread.
+   */
+  public static void unsetMetricsContainer() {
+    CONTAINER_FOR_THREAD.remove();
+  }
+
+  /**
+   * Return the {@link MetricsContainer} for the current thread.
+   */
+  public static MetricsContainer getCurrentContainer() {
+    MetricsContainer container = CONTAINER_FOR_THREAD.get();
+    if (container == null) {
+      throw new IllegalStateException("Must call setMetricsContainer before reporting metrics.");
+    }
+    return container;
+  }
+
+  /**
+   * Return the {@link CounterCell} that should be used for implementing the given
+   * {@code metricName} in this container.
+   */
+  public CounterCell getOrCreateCounter(MetricName metricName) {
+    return counters.getOrCreate(metricName);
+  }
+
+  public DistributionCell getOrCreateDistribution(MetricName metricName) {
+    return distributions.getOrCreate(metricName);
+  }
+
+  private <UpdateT, CellT extends MetricCell<UpdateT>>
+  ImmutableList<MetricUpdate<UpdateT>> extractUpdates(
+      MetricsMap<MetricName, CellT> cells,
+      boolean includeZero) {
+    ImmutableList.Builder<MetricUpdate<UpdateT>> updates = ImmutableList.builder();
+    for (Map.Entry<MetricName, CellT> cell : cells.entries()) {
+      UpdateT update = cell.getValue().getDeltaUpdate(includeZero);
+      if (update != null) {
+        updates.add(MetricUpdate.create(MetricKey.create(stepName, cell.getKey()), update));
+      }
+    }
+    return updates.build();
+  }
+
+  /**
+   * Return the delta updates for all metrics in this container.
+   */
+  public MetricUpdates getDeltas() {
+    boolean includeZero = !committedFirstDelta.get();
+    return MetricUpdates.create(
+        extractUpdates(counters, includeZero),
+        extractUpdates(distributions, includeZero));
+  }
+
+  private <UpdateT, CellT extends MetricCell<UpdateT>>
+  void commitDeltas(MetricsMap<MetricName, CellT> cells, Iterable<MetricUpdate<UpdateT>> updates) {
+    for (MetricUpdate<UpdateT> counterUpdate : updates) {
+      CellT cell = checkNotNull(cells.get(counterUpdate.getKey().metricName()),
+          "Metric cell %s in delta being committed should have already been defined.");
+      cell.commitDeltaUpdate(counterUpdate.getUpdate());
+    }
+  }
+
+  /**
+   * Record all of the given deltas as reported.
+   *
+   * <p>The updates in {@code deltas} must have been produced by calling {@link #getDeltas()} on
+   * this metrics container, and no deltas should have been committed since the associated call to
+   * {@link #getDeltas()}.
+   */
+  public void commitDeltas(MetricUpdates deltas) {
+    commitDeltas(counters, deltas.counterUpdates());
```

Comment:
Delta-based metrics will always be lossy. Rather than changing this API, we should just add tests to make sure this doesn't fail part-way through. But other problems (machine/network failure) can still happen causing us to drop updates.

An alternative would be to always send cumulative values, and just report those periodically (when values have changed). The difficulty here is that the runner then needs to figure out how to exclude the previous cumulative value to prevent double-counting.

If we go this route, then we can just use a simple dirty-state on the metric to figure out what we need to send. Specifically, if it has the values CLEAN, COMMITTING and DIRTY, we can make a best effort to avoid sending CLEAN metrics as follows:

Before reporting updates: {DIRTY, COMMITTING} -> COMMITTING (ignore CLEAN)
After reporting updates: COMMITTING -> CLEAN (ignore DIRTY and CLEAN)
Whenever metric changes: {...} -> DIRTY

This ensures that a metric is only CLEAN if we've reported all updates for it and does a best-effort attempt at only reporting DIRTY metrics.

The more I think about it the more I lean back towards though, even though it does complicate the runner a little -- thoughts?